src/OpenLoco/src/Vehicles/Vehicle1.cpp

```diff
@@ -126,13 +130,167 @@ namespace OpenLoco::Vehicles
+
+        const auto height = World::TileManager::getHeight(veh2.position);
+        const auto volume = veh2.position.z < height.landHeight ? -1500 : 0;
+
+        Audio::playSound(Audio::makeObjectSoundId(soundObjId), veh2.position + World::Pos3{ 0, 0, 22 }, volume, 22050);
+    }
+
     // 0x004A97A6
     bool Vehicle1::updateRail()
     {
-        registers regs;
-        regs.esi = X86Pointer(this);
+        Vehicle train{ head };
+        uint16_t curveSpeedFraction = std::numeric_limits<uint16_t>::max();
+        Speed16 maxSpeed = kSpeed16Max;
+        bool isOnRackRail = false;
+        if (!train.head->hasVehicleFlags(VehicleFlags::manualControl))
+        {
+            RoutingManager::RingView ring(routingHandle);
+            for (auto iter = ring.rbegin(); iter != ring.rend(); ++iter)
+            {
+                auto res = RoutingManager::getRouting(*iter);
+                isOnRackRail |= (res & (1U << 13)) != 0; // rackrail
+                uint8_t trackId = (res >> 3) & 0x3F;
+                curveSpeedFraction = std::min(curveSpeedFraction, _trackIdToCurveSpeedFraction[trackId]);
+                if (res & (1U << 12))
+                {
+                    const auto* bridgeObj = ObjectManager::get<BridgeObject>((res & 0xE00) >> 9);
+                    if (bridgeObj->maxSpeed != kSpeed16Null)
+                    {
+                        maxSpeed = std::min(bridgeObj->maxSpeed, maxSpeed);
+                    }
+                }
+            }
+        }
+
+        const auto* trackObj = ObjectManager::get<TrackObject>(trackType);
+        const Speed32 fractionalSpeed = Speed32(static_cast<uint32_t>(curveSpeedFraction) * trackObj->curveSpeed.getRaw());
+        maxSpeed = std::min(toSpeed16(fractionalSpeed + 1.0_mph), maxSpeed);
+
+        if (train.head->has38Flags(Flags38::unk_5))
+        {
+            maxSpeed += maxSpeed / 4;
+            maxSpeed = std::min(trackObj->curveSpeed, maxSpeed);
+        }
+
+        const auto veh2MaxSpeed = [veh2 = train.veh2]() {
+            if (veh2->has73Flags(Flags73::isBrokenDown))
+            {
+                return veh2->maxSpeed / 4;
+            }
+            return veh2->maxSpeed;
+        }();
 
-        return !(call(0x004A97A6, regs) & X86_FLAG_CARRY);
+        maxSpeed = std::min(maxSpeed, veh2MaxSpeed);
+        if (isOnRackRail)
+        {
+            maxSpeed = std::min(maxSpeed, train.veh2->rackRailMaxSpeed);
+        }
+
+        if (!train.head->hasVehicleFlags(VehicleFlags::manualControl))
+        {
+            maxSpeed = std::min(maxSpeed, Speed16(var_3C / 32768) + 5_mph);
+        }
+
+        if ((train.head->hasVehicleFlags(VehicleFlags::manualControl) && train.head->var_6E <= -20)
+            || train.head->hasVehicleFlags(VehicleFlags::commandStop))
+        {
+            if (train.veh2->currentSpeed == 0.0_mph)
+            {
+                maxSpeed = 0_mph;
+            }
+        }
+        var_44 = maxSpeed;
```

Comment:
names for any of these? `var_44`, `var_3C`, `unk2`, `_vehicleUpdate_var_1136134` and the flags for `has73Flags` and `has38Flags`?

Reply:
I think i've give names to everything i'm confident in now.